torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp

```diff
@@ -839,6 +846,17 @@ class TORCH_API ProcessGroupNCCL : public Backend {
 
   bool storeError_{false};
 
+  // The lock which protects the write/read of extendedTimeout_
+  std::mutex mtxTimeoutExtension_;
+
+  // The extended timeout on top of existing timeout for works issued before
+  // first work finishes.
+  std::chrono::milliseconds extendedTimeout_ = std::chrono::milliseconds(0);
+
+  // Record of first work and will get reset if timeout is further extended.
+  c10::intrusive_ptr<ProcessGroupNCCL::WorkNCCL> firstWorkSinceExtended_ =
+      nullptr;
```

Comment:
My comment from the Phabricator diff still applies here https://www.internalfb.com/diff/D60274726?transaction_fbid=1141409296951808

> You don't need an entire queue. Let's assume that NCCL collectives complete in the order they were enqueued. You need only two quantities: "accumulated ephemeral timeout" (total_timeout) and "timeout pending reset" (inflight_timeout). Whenever you extend the timeout, you increase the accumulated timeout. Whenever you issue a collective, you record timeout to reset (work_timeout_reduce) after this collective completes as "total_timeout - inflight_timeout" on the Work object, and then set inflight_timeout := total_timeout. The invariant is that the sum of all work_timeout_reduce on outstanding Work objects is equal to inflight_timeout. When the work completes, you reduce total_timeout and inflight_timeout by the work_timeout_reduce. Eventually, you will quiesce back to zero total_timeout.

In particular, I do not think you should need to store the first work since extended. It seems more logical to associate the timeout pending reset ON the work object itself, this allows for multiple pending work with different resets.

Reply:
This proposal also works and I think I get it. But why would a Work object care about `work_timeout_reduce` it's essentially a concept outside Work? What each work wants to know is what is my own timeout, IMHO.

Reply:
Actually we draw this one to ensure we understand you correctly:

<img width="1173" alt="image" src="https://github.com/user-attachments/assets/e9e64688-edf8-497b-bb0d-4f31759e3fd1">


Yeah this indeed is better than the current one because if there are interleave of collective finishing and timeout extension, we don't bump up timeout monotonically. But instead of storing it into Work object, we want to use a map to store.

Reply:
Yup, your table looks right to me.

If you want to argue about storing it on Work or storing it in a map, well, it's all semantically isomorphic in the end so I don't really care. I had been coming at it from the perspective that Work was a specifically c10d collective concept though. Note that if you do a map you have to worry about (1) lifetimes, (2) ABA problems, (3) it's a little less efficient because you have to do the lookup, (4) locking.

Reply:
Got it. I agree, we need to handle the lifetimes for the map and it has extra cost.. But we will need a lock anyway even if we store the reduce_time into work object and with lock, there shouldn't be an ABA problem because each time, read/write happens inside a mutex which makes it atomic?

Reply:
I'm definitely not saying the other problems aren't solvable, just need to make sure the i's are dotted and the t's crossed